packages/api/src/apps/api/app.ts

```diff
@@ -97,25 +98,46 @@ async function run(env: ProcessEnv) {
   // expose calls through express
   await Express({ port: Number(env.EXPRESS_PORT) }, actions);
 
-  // main update loop, update every block
-  provider.on("block", (blockNumber: number) => {
+  // break all state updates by block events into a cleaner function
+  async function updateByBlock(blockNumber: number) {
```

Comment:
One thing I realized about this architecture (that I didn't realize before) is that it seems that we're designing this for a single process that does both ingestion and exposes the API. This seems to be specifically designed for a single-process and not generalizable for a multi-process system backed by a DB (which is our long term vision if I understand correctly).

Do you think it makes sense to separate out the API piece and the ingestion piece into distinct files/functions? This would allow us to run each by themselves and _also_ have a file that pulls both in and runs them together for an in-memory store. We could also do this all in one file and allow this to be customized via CLI params or env variables, but what's your vision for how we can run these parts separately?

Reply:
i think this lends itself quite well to a multi process shared db architecture, with some minor updates. Assume the "appState" is based all on shared DB tables, each service only depends on these tables for state, some write to them, others read. This means that you can break out each service into its own application, inject those same tables and thing should behave identically to the monolith.

Reply:
Totally agree that the lower-level components are agnostic. I was more talking about this file. Definitely something to worry about in a later PR, but it might be nice to have a way to compose the trigger, update process, and query handler so you can mix and match since those are things we probably want to vary across different environments. 

This could be especially cool for spinning up a local instance to test against in dapps. The dapps could even spin something like this up in process!

Reply:
cool, just to touch on this point and your other comment, i see the app files as bespoke, ie they are opinioned/custom compositions of lower level components. There's a couple schools of though on adding further features, one is to add switches within the app component with configuration through params (envs) to enable/disable features.  The other would be to create a completely new app for each type of feature. In practice i tend to lean towards the latter approach, though this depends on what needs to be added.

For example, we may want a serverless style of ingestion to db. Rather than have a serverless mode in the existing app, just make a new app which does this and call it `serverless-updater` or something.  This is mainly just a style thing but there are benefits as i find its less risky than changing working code, and the separation keep code simpler in the long run.

Reply:
(These comments are not blockers for merge, so please feel free to merge whenever!)

Agreed, you could definitely build multiple apps or have one app that was customized. I assumed you preferred the latter since this one is named app.ts (implying that there's only one).

The main point I'm making is that it seems like there's less modularity at this level than at the lower levels, but this is a level that we know we want to generalize. There's some distinct code in here that's clearly a class of "ingester" that handles grabbing new data. There's another set of distinct code that is involved in serving data using actions. Lastly, there seems to be a global state object initialized in-line and is a storage layer for state.

These all feel like distinct responsibilities to me, but I don't have super strong opinions. I just imagine we'll end up rewriting much of this code in other apps.

Reply:
(These comments are not blockers for merge, so please feel free to merge whenever!)

Agreed, you could definitely build multiple apps or have one app that was customized. I assumed you preferred the latter since this one is named app.ts (implying that there's only one).

The main point I'm making is that it seems like there's less modularity at this level than at the lower levels, but this is a level that we know we want to generalize. There's some distinct code in here that's clearly a class of "ingester" that handles grabbing new data. There's another set of distinct code that is involved in serving data using actions. Lastly, there seems to be a global state object initialized in-line and is a storage layer for state.

These all feel like distinct responsibilities to me, but I don't have super strong opinions. I just imagine we'll end up rewriting much of this code in other apps.

Reply:
(These comments are not blockers for merge, so please feel free to merge whenever!)

Agreed, you could definitely build multiple apps or have one app that was customized. I assumed you preferred the latter since this one is named app.ts (implying that there's only one).

The main point I'm making is that it seems like there's less modularity at this level than at the lower levels, but this is a level that we know we want to generalize. There's some distinct code in here that's clearly a class of "ingester" that handles grabbing new data. There's another set of distinct code that is involved in serving data using actions. Lastly, there seems to be a global state object initialized in-line and is a storage layer for state.

These all feel like distinct responsibilities to me, but I don't have super strong opinions. I just imagine we'll end up rewriting much of this code in other apps.